temporalio/worker/interceptor.py

```diff
@@ -90,3 +114,191 @@ def info(self) -> temporalio.activity.Info:
+    run_timeout: Optional[timedelta]
+    task_timeout: Optional[timedelta]
+    id_reuse_policy: temporalio.common.WorkflowIDReusePolicy
+    retry_policy: Optional[temporalio.common.RetryPolicy]
+    cron_schedule: str
+    memo: Optional[Mapping[str, Any]]
+    search_attributes: Optional[Mapping[str, Any]]
+    # The types may be absent
+    arg_types: Optional[List[Type]]
+    ret_type: Optional[Type]
+
+
+@dataclass
+class StartLocalActivityInput:
+    """Input for :py:meth:`WorkflowOutboundInterceptor.start_local_activity`."""
+
+    activity: str
+    args: Iterable[Any]
+    activity_id: Optional[str]
+    schedule_to_close_timeout: Optional[timedelta]
+    schedule_to_start_timeout: Optional[timedelta]
+    start_to_close_timeout: Optional[timedelta]
+    retry_policy: Optional[temporalio.common.RetryPolicy]
+    local_retry_threshold: Optional[timedelta]
+    cancellation_type: temporalio.workflow.ActivityCancellationType
+    # The types may be absent
+    arg_types: Optional[List[Type]]
+    ret_type: Optional[Type]
+
+
+class WorkflowInboundInterceptor:
+    """Inbound interceptor to wrap outbound creation, workflow execution, and
+    signal/query handling.
+
+    This should be extended by any workflow inbound interceptors.
+    """
+
+    def __init__(self, next: WorkflowInboundInterceptor) -> None:
+        """Create the inbound interceptor.
+
+        Args:
+            next: The next interceptor in the chain. The default implementation
+                of all calls is to delegate to the next interceptor.
+        """
+        self.next = next
+
+    def init(self, outbound: WorkflowOutboundInterceptor) -> None:
+        """Initialize with an outbound interceptor.
+
+        To add a custom outbound interceptor, wrap the given interceptor before
+        sending to the next ``init`` call.
+        """
+        self.next.init(outbound)
+
+    async def execute_workflow(self, input: ExecuteWorkflowInput) -> Any:
+        """Called to run the workflow."""
+        return await self.next.execute_workflow(input)
+
+    async def handle_signal(self, input: HandleSignalInput) -> None:
+        """Called to handle a signal."""
+        return await self.next.handle_signal(input)
+
+    async def handle_query(self, input: HandleQueryInput) -> Any:
+        """Called to handle a query."""
+        return await self.next.handle_query(input)
+
+
+class WorkflowOutboundInterceptor:
+    """Outbound interceptor to wrap calls made from within workflows.
+
+    This should be extended by any workflow outbound interceptors.
+    """
+
+    def __init__(self, next: WorkflowOutboundInterceptor) -> None:
+        """Create the outbound interceptor.
+
+        Args:
+            next: The next interceptor in the chain. The default implementation
+                of all calls is to delegate to the next interceptor.
+        """
+        self.next = next
+
+    async def continue_as_new(self, input: ContinueAsNewInput) -> NoReturn:
```

Comment:
So this ended up being async? Is the still required after our last discussion on slack?

Reply:
I am in the process of changing this now. This will become non-async and throw and only if the exception is bubbled all the way out of the workflow will we return continue as new.

I was still struggling with whether I'd make the error public and/or if I should extend from `BaseException` instead of `Exception` to discourage catching it. For now I'll make it a public error that extends from `BaseException`. Can change if needed.

Reply:
I have updated this.